### PR TITLE
Add inlining optimisation.

### PR DESCRIPTION
**Pull request by pbrunet**
* Usefull to propagate more expression template.

I did some test with

```
#pythran export bar(float[])                                                     
import numpy                                                                     
                                                                                 
def foo(a):                                                                      
    return numpy.cos(a)                                                          
                                                                                 
def bar(c):                                                                      
    return foo(c) + c * foo(c)  
```

and I am 10% faster. (On my slow computer...)

### CODE REVIEW

File: pythran/optimizations/inlining.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            * It is a one line function
+            * It is not recursive
+        """
+        for stmt in node.body:
+            if isinstance(stmt, ast.FunctionDef):
+                if len(stmt.body) == 1:
+                    ids = self.passmanager.gather(Identifiers, stmt.body[0],
+                                                  self.ctx)
+                    if stmt.name not in ids:
+                        self.fun[stmt.name] = stmt
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: In the future, the notion of what is inlinable and what is not may evolve and grow more complex. I'll advocate for an Inlinable analysis separated from this transformation.

File: pythran/optimizations/inlining.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        return self.generic_visit(node)
+
+    def visit_Call(self, node):
+        """ Replace function call by inlined function's body. """
+        if isinstance(node.func, ast.Name) and node.func.id in self.fun:
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: no no no, you should rely on aliasing, and not name id, to perform inlining. remember an identifier name has nothing to do with is actual value :-)

File: pythran/optimizations/inlining.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            self.update = True
+            to_inline = copy.deepcopy(self.fun[node.func.id])
+            arg_to_value = {a.id: b
+                            for a, b in zip(to_inline.args.args, node.args)}
+            return Inlineur(arg_to_value).visit(to_inline.body[0])
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: what a frenchism « Inlineur » → Inliner ?

File: pythran/optimizations/inlining.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class Inlineur(ast.NodeTransformer):
+
+    """ Helper transform that perfomed inlined body transformation. """
+
+    def __init__(self, match):
+        """ match : {original_variable_name : Arguments use on call}. """
+        self.match = match
+        super(Inlineur, self).__init__()
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: Ther's one additional issue here : what if the argument is used twice, and the effective argument is an expression , You should create a temporary variable and let forward substituion do the job

File: pythran/pythran.cfg [hunk 5]
```diff
@@ -11,6 +11,7 @@ optimizations = pythran.optimizations.ForwardSubstitution
                 pythran.optimizations.PatternTransform
                 pythran.optimizations.Square
                 pythran.optimizations.RangeLoopUnfolding
+                pythran.optimizations.Inlining
```
Comment (serge-sans-paille) [inline pythran/pythran.cfg]: Maybe put it before forwrad substitution?

**Comment by serge-sans-paille** [comment]
Great proposal! Inlining is a complex subject and you fell in a few booby traps ;-)
You're lacking a few test cases too?

**Comment by serge-sans-paille** [comment]
btw, I foresee a cool article on this PR :-)

File: pythran/optimizations/inlining.py (hunk 2 above)
Comment (pbrunet) [inline pythran/optimizations/inlining.py]: I am pretty rusty.... :-)

**Comment by pbrunet** [comment]
I am not sure I need more tests as recursion is already tested with many others cases and the basic transformation is tested with doctest.

I do the suggested modiciation but sadly ForwardSubstition doesn't really help here :s 

```
#pythran export bar(float[])                                                     
import numpy                                                                     

def foo(a):                                                                      
    return numpy.cos(a)                                                          

def bar(c):                                                                      
    return foo(c) + c * foo(c)  
```

In this case, bar will become  : 
```
def bar(c):
    _pythran_fooa0 = c
    _pythran_fooa1 = c
    return numpy.cos(_pythran_fooa0) + c * numpy.cos(_pythran_fooa1) 
```

ForwardSubstition can't substitute here as pythran_fooa0 alias on `c` so it is use 4 times (no FS if it is use more than once)
May be we should improve FS to foward with more than one "use" in the case of "simple variable assignment"?

**Comment by serge-sans-paille** [comment]
I reply point-by-point


> I am not sure I need more tests as recursion is already tested with many others cases and the basic transformation is tested with doctest.

Right now your tests are very conservative. We may want to extend them
later on (eg to support more complex inlining situations)

> I do the suggested modiciation but sadly ForwardSubstition doesn't really help here :s 
> 
> ```
> #pythran export bar(float[])                                                     
> import numpy                                                                     
> 
> def foo(a):                                                                      
>     return numpy.cos(a)                                                          
> 
> def bar(c):                                                                      
>     return foo(c) + c * foo(c)  
> ```
> 
> In this case, bar will become  : 
> ```
> def bar(c):
>     _pythran_fooa0 = c
>     _pythran_fooa1 = c
>     return numpy.cos(_pythran_fooa0) + c * numpy.cos(_pythran_fooa1) 
> ```


The primary goal was to better handle

```
def foo(a):
    return coas(a) + sin(a)
```

when the call site is

```
foo(b +c)
```

And your proposal should already improve the situation.

> ForwardSubstition can't substitute here as pythran_fooa0 alias on `c` so it is use 4 times (no FS if it is use more than once)
> May be we should improve FS to foward with more than one "use" in the case of "simple variable assignment"?

Isn't this an error in our FS implementation? I see only one use of _pythran_fooa0...

**Comment by pbrunet** [comment]
Indead it is not an implementation error but Lazyness analysis was done to
determine "lazy" types. In the given case, if pythran_fooa0 is a lazy type,
as it alias on c, c is a lazy type and they will be evaluated 4 times (it
is bad for lazy type like numpy_expr). So we assume that it is not lazy.
But it is based on lazy type which is stricter than FS. We should certainly
looks only for variable name instead of aliasing in this case.

2016-03-22 20:19 GMT+01:00 serge-sans-paille <notifications@github.com>:

File: pythran/analyses/inlinable.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    def __init__(self):
+        self.result = dict()
+        super(Inlinable, self).__init__()
+
+    def visit_FunctionDef(self, node):
+        """ Determine this function definition can be inlined. """
+        if (len(node.body) == 1 and
+                isinstance(node.body[0], (ast.Call, ast.Return))):
+            ids = self.passmanager.gather(Identifiers, node.body[0], self.ctx)
```
Comment (serge-sans-paille) [inline pythran/analyses/inlinable.py]: This does not work in the (strange) case where
```
def foo(foo):
    return foo
```

File: pythran/optimizations/inlining.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __init__(self):
+        """ fun : Function {name :body} for inlinable functions. """
+        self.update = False
+        self.defs = list()
+        self.call_count = 0
+        super(Inlining, self).__init__(Inlinable, Aliases)
+
+    def visit_Stmt(self, node):
+        """ Add new variable definision before the Statement. """
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: defini*t*ion

File: pythran/optimizations/inlining.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    visit_Return = visit_Stmt
+    visit_Assign = visit_Stmt
+    visit_AugAssign = visit_Stmt
+    visit_Print = visit_Stmt
+    visit_For = visit_Stmt
+    visit_While = visit_Stmt
+    visit_If = visit_Stmt
+    visit_With = visit_Stmt
+    visit_Assert = visit_Stmt
+    visit_Expr = visit_Stmt
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: you could write `visit_Return = visit_Assign = ... = visit_Stmt`

File: pythran/optimizations/inlining.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    def visit_Call(self, node):
+        """
+        Replace function call by inlined function's body.
+
+        We can inline if it alias on only one function.
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: alias*es*

File: pythran/optimizations/inlining.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        """
+        func_aliases = self.aliases[node.func].aliases
+        if len(func_aliases) == 1:
+            function_def = iter(func_aliases).next()
+            if (isinstance(function_def, ast.FunctionDef) and
+                    function_def.name in self.inlinable):
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: is the first check meaningful? i mean if we get, say, NewMem, it cannot be in self.inlinable, right? Add an assert if you're not sure !

File: pythran/optimizations/inlining.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                self.update = True
+                to_inline = copy.deepcopy(self.inlinable[function_def.name])
+                arg_to_value = dict()
+                values = node.args
+                values += to_inline.args.defaults[len(node.args) -
+                                                  len(to_inline.args.args):]
+                for arg_fun, arg_call in zip(to_inline.args.args, values):
+                    v_name = "_pythran_inline_{}{}{}".format(function_def.name,
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: we generally use double underscore as a prefix, as it's *reserved for the compiler*

File: pythran/optimizations/inlining.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __init__(self, match):
+        """ match : {original_variable_name : Arguments use on call}. """
+        self.match = match
+        super(Inliner, self).__init__()
+
+    def visit_Name(self, node):
+        """ Transform name from match values if available. """
+        if node.id in self.match:
+            return self.match[node.id]
```
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: `return self.match.get(node.id, node)`

File: pythran/analyses/inlinable.py (hunk 6 above)
Comment (pbrunet) [inline pythran/analyses/inlinable.py]: Yes, but it is still "safe". Should I check for name not in identifiers except if in args? Does it looks enough to you?

File: pythran/optimizations/inlining.py (hunk 8 above)
Comment (pbrunet) [inline pythran/optimizations/inlining.py]: I could but I am not sure it is "better", is it?

File: pythran/optimizations/inlining.py (hunk 10 above)
Comment (pbrunet) [inline pythran/optimizations/inlining.py]: My code say : If you have only one alias and it is inlinable , then... and you say : OK but with one alias, you can get not inlinable. It is true, that's why I check both. If you have multiple alias and all of them are inlinable, you can't inline. If you have one alias but not inlinable, you can't inline.

File: pythran/optimizations/inlining.py (hunk 12 above)
Comment (pbrunet) [inline pythran/optimizations/inlining.py]: yeah !!

File: pythran/analyses/inlinable.py (hunk 6 above)
Comment (serge-sans-paille) [inline pythran/analyses/inlinable.py]: Yes your solution is safe, just add a FIXME ;-)

File: pythran/optimizations/inlining.py (hunk 8 above)
Comment (serge-sans-paille) [inline pythran/optimizations/inlining.py]: I don't know. Do as you want!

**Comment by serge-sans-paille** [comment]
@pbrunet still a test issue :-/

**Pull request merged**